Join stake pool endpoint (part 1 - no signCert impl)

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#896 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added pools check in endpoint
- [x] I have added createCert
- [x] I have added submitCert 
- [x] I have added computeCertFee 
- [x] I have added test and adapted other tests affected 

# Comments

<!-- Additional comments or screenshots to attach if any -->
In the forthcoming PR I will tackle signCert impl (waiting here for mkDelegationCertTx to find its way into master first)
<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
handle NoSuchPool error in joinStakePools

paweljakubas committed:
add first part of deriveDelegationAccount in Cardano.Wallet with retrieving root key errors

extend constraints in pools and prepare to use withRootKey

use withRootKey to get chimeric account private key

move logic to Cardano.Wallet and add coin selection part

adjusting for fee

use withWorkerCtx to embrace DBLayer constraint

change from Hardened to Soft

paweljakubas committed:
add createCert and resign from coin selection

paweljakubas committed:
add computeCertFee and generalize feeOpt in Cardano.Wallet

paweljakubas committed:
change error ErrJoinStakePool to ErrCreateCert for consistency

add incomplete signCert in Cardano.Wallet

use signCert in joinStakePool handle and add LiftHandle instances of ErrSignCert

paweljakubas commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +983,97 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+            (tx, wit) <- either
+                (throwE . ErrSignCert)
+                pure
+                (mkCertificateTx tl keyFrom (poolId, addrPrv, pwd) ins allShuffledOuts)
```
waiting for `mkCertificateTx` PRs to be merged

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -786,15 +800,47 @@ listPools spl =
+                W.signCert @_ @s @t @k wrk wid () pwd selection poolId
+
+            liftHandler $ withWorkerCtx ctx wid liftE3 $ \wrk ->
+                W.submitTx @_ @s @t @k wrk wid (tx, meta, wit)
```
I need to think once again if submitTx can be used here, especially in the context of its interaction to DB

paweljakubas committed:
add submitCert function

paweljakubas committed:
use updateWallet in joinStakePool

paweljakubas committed:
add test for NoSuchPool error

paweljakubas commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +985,119 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+    mapExceptT atomically $ withExceptT ErrSubmitCertNoSuchWallet $ do
+        putTxHistory (PrimaryKey wid) [(tx, meta)]
+        let (TxMeta _ _ sl _ _) = meta
+        putDelegationCertificate (PrimaryKey wid) poolId sl
```
writing both tx history and delegation certificate in DB

KtorZ replied:
:ok_hand: 

suggestion:

```hs
putDelegationCertificate (PrimaryKey wid) poolId (meta ^. #slotId)
```

why ? Avoiding pattern-maching here makes for code that is slightly easier to maintain. Adding fields to `TxMeta` won't require us to go back to this very point to add an extra `_`. Alternatively, using `NamedFieldPuns`, I'd also favor the following approach:

```hs
let TxMeta{sl} = meta
putDelegationCertificate (PrimaryKey wid) poolId sl
```

I'd still favor the first one because the entire line is then captured in a single expression and is easier to load in brain :)

paweljakubas replied:
opted for lens version

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -786,15 +801,52 @@ listPools spl =
+            liftHandler $ withWorkerCtx ctx wid liftE3 $ \wrk ->
+                W.submitCert @_ @s @t @k wrk wid poolId (tx, meta, wit)
+
+            liftHandler $ withWorkerCtx ctx wid throwE $ \wrk ->
```
changing wallet metadata. now delegating with pool id. wonder if we need to check first if wallet is already not delegating. if yes, then error is thrown as we need to first quickStakePool. I have heard two versions, this one. and also that joinStakePool just changes this field. which means changes the stake pool it delegates. Another question. what to do if we are already joined to the pool and want to rejoin to the same one. I am tentatively in favor of also throwing error. Any thought/suggestion here?

KtorZ replied:
Ah! But this is actually slightly _wrong_. We haven't joined the stake pool at the very stage and aren't yet delegating. 
We have only _made a request_ about it. The delegation will only be effective in the epoch following the insertion of the certificate in the ledger. Detecting that we are delegating is a job for #899 ! So here, there's no need to update the wallet metadata, they'll be updated when the certificate is discovered.

paweljakubas replied:
thanks for this remark. removed this. hope we will tackle this in #899

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/Fee.hs:
```diff
@@ -93,6 +94,24 @@ computeFee policy (Quantity sz) =
+    -> Quantity "byte" Int
+    -> Fee
+computeCertFee policy (Quantity sz) =
+    Fee $ ceiling (a + b*fromIntegral sz + c)
```
I am here assuming that this is flat fee for certificates and I can delegate to one stake pool. If this assumption is not true, then I should change the signature of this function and add `number of certificates`. On the other hand I see, something like this : https://slack-redir.net/link?url=https%3A%2F%2Fgithub.com%2Finput-output-hk%2Fchain-libs%2Fblob%2Fmaster%2Fchain-impl-mockchain%2Fdoc%2Fformat.md%23type-0-initial-blockchain-configuration and here 
```
LinearFee has the format:

Constant | Coefficient | Certificate
all of them 64-bit unsigned integers, specifying how fees are computed using the formula:

Constant + Coefficient * (inputs + outputs) + Certificate * certificates
where inputs, outputs and certificates represent the size of the serialization of the corresponding parts of a transaction in bytes.
```
so according to this document `certificates` is size of serialization part

KtorZ replied:
I agree that this function should probably take an extra argument to reflect the number of certificates. And, at the call-site, use a value of `1` since this is where we actually _know_ that we are only creating a single certificate.

paweljakubas commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -153,6 +161,20 @@ spec = do
 
+    it "STAKE_POOLS_JOIN_01 - Cannot join non-existant stakepool" $ \ctx -> do
+        let pId =
+                "4f8d686a02c6e625b5a59cc9e234f32e5d72987012f9c25c9a6b60ddade197d9" :: Text
```
I know that I need to use `pIdAbsent` here in the end. I just get here 
```
\"4f8d686a02c6e625b5a59cc9e234f32e5d72987012f9c25c9a6b60ddade197d9\"
```
when using `show pIdAbsent`

KtorZ replied:
Alternatively, use `0000000.....00000` for an unknown stake pool. Chances of getting a stake pool with such id are null (1 in 2^32 exactly)

paweljakubas replied:
used 00000....00000 and refactored the test case a little bit

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +985,119 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+    let sel = CoinSelection [] [] []
+    withExceptT ErrCreateCertFee $ do
+        debug tr "Coins selected for delegation certificate after fee adjustment"
+            =<< adjustForFee (feeOpts tl computeCertFee (bp ^. #getFeePolicy)) utxo sel
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +985,119 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+    -> PoolId
+    -> ExceptT ErrSignCert IO (Tx, TxMeta, UTCTime, [TxWitness])
+signCert ctx wid argGenChange pwd
+    (CoinSelection ins outs chgs) _poolId = db & \DBLayer{..} -> do
```
style sugggestion: 

```hs
signCert ctx wid argGenChange pwd coinSel _poolId = db & \DBLayer{..} -> do
    let (CoinSelection ins outs chgs) = coinSel
    ...
```

paweljakubas replied:
done

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +985,119 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+                    forM chgs $ \c -> do
+                        addr <- state (genChange argGenChange)
+                        return $ TxOut addr c
+            allShuffledOuts <- liftIO $ shuffle (outs ++ changeOuts)
```
The shuffling is so-to-speak pointless in this scenario because all outputs are already ours. In the case of standard transaction, we shuffle to add some privacy and avoid having outputs that follow a very specific pattern:

- recipient
- change
- recipient
- change
etc ...

Here, there's no need :)

paweljakubas replied:
indeed, corrected

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +985,119 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+                (throwE . ErrSignCert)
+                pure
+                (mkCertificateTx tl keyFrom (poolId, addrPrv, pwd) ins allShuffledOuts)
+            --}
```
Generally speaking, I'd suggest putting an explicit `error "not implemented"` and avoid writing functions that:

1. Compiles
2. Raises no runtime exception
3. Are not correct

In this case, I am pretty sure that we won't forgot to update it, but, as a habit, it's good to make sure that "if it compiles, it works".  If it doesn't work, either make sure it doesn't compile, or make sure it fails loudly :)

paweljakubas replied:
my intention was to wait for mkCertificateTx to be merged, then rebase this PR and only when the functionalities work as we expect and all integration tests pass, merge this PR (according to the rule, that only well tested changes can be merged, no partial, untested code in master etc)

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +985,119 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+                    }
+            let time = slotStartTime
+                    (slotParams (blockchainParameters cp))
+                    txSlot
```
:+1: I see room for code factorization with the `signTx` function that is very analogous. I wouldn't recommend doing it in this PR, but definitely leave a bullet point in the development plan of the ticket about refactoring this.

paweljakubas replied:
added a point in development plan : 
```
[ ] Consider reusing create/sign/submit parts between Tx and Cert
```

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1579,6 +1631,55 @@ instance LiftHandler ErrMetricsInconsistency where
+                    , "uses one or more of the same inputs, or it may be due "
+                    , "to some other reason. Here's an error message that may "
+                    , "help with debugging: ", pretty err
+                    ]
```
Can't we re-use previous handlers here?

paweljakubas committed:
comply with the first review of draft PR

paweljakubas committed:
add ErrCreateCertPoolAlreadyJoined

slim down the PR and omit signCert plus adjust integration tests

adjust other tests

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +982,70 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+    withExceptT ErrSubmitCertNetwork $ postTx nw binary
+    mapExceptT atomically $ withExceptT ErrSubmitCertNoSuchWallet $ do
+        putTxHistory (PrimaryKey wid) [(tx, meta)]
+        putDelegationCertificate (PrimaryKey wid) poolId (meta ^. #slotId)
```
**[change required]**
This line should be removed. 

**Why**
We don't keep track of "pending delegation certificates" the same way we keep track of transaction. So, until a certificate made it into the ledger (and is therefore, discover by the restoration loop), it simply doesn't exist. In particular, the "slotId" would be quite wrong here; it should refer to the slot at which the certificate is inserted, not the one at which we made the request (which has very little meaning from the blockchain PoV).

paweljakubas replied:
done

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -786,15 +801,53 @@ listPools spl =
+        Nothing ->
+            liftHandler $ throwE (ErrCreateCertNoSuchPool poolId)
+        Just _ -> do
+            let (ApiWalletPassphrase (ApiT pwd)) = passwd
```
**[suggestion: readability]**

I'd suggest to move the guard in a dedicated function (possibly declared in the where clause) to have something which is more consistent with the style adopted so far:

```hs
whenPoolIsUnknown poolId  $ liftHandler $ throwE (ErrCreateCertNoSuchPool poolId)
```

**Why**

This keeps the `joinStakePool` function more readable by moving some lower-level concerns into a separate function. Then, `joinStakePool` reads more like a sentence.

paweljakubas replied:
done

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -786,15 +801,53 @@ listPools spl =
+            selection <- liftHandler $ withWorkerCtx ctx wid liftE1 $ \wrk ->
+                W.createCert @_ @s @t wrk wid
+
+            _ <- throwError err501
```
**[suggestion: potentially error prone]**
Remove this line.

**Why**
`signCert` will already throw loudly and cause the request to fail with a 500. So this looks redundant to me. Otherwise, we need to remember to modify this section of the code after implementing `signCert`. This could be confusing, although, this is very minor.

paweljakubas replied:
indeed, done

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1579,6 +1635,43 @@ instance LiftHandler ErrMetricsInconsistency where
+                , toText poolId
+                ]
+        ErrCreateCertPoolAlreadyJoined poolId ->
+            apiError err404 PoolAlreadyJoined $ mconcat
```
**[change required]**
This should rather be `err403`

**Why**
`404` is used for missing pages or resources. Here, the case is slightly different: the pool exists, the request is syntactically valid, but we refuse to proceed. That's the semantic given to `403 Forbidden`

paweljakubas replied:
corrected

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -126,39 +133,51 @@ spec = do
-        expectResponseCode HTTP.status501 r2
+        expectResponseCode HTTP.status403 r2
+
+    it "STAKE_POOLS_JOIN_01 - Cannot join non-existant stakepool" $ \ctx -> do
```
:+1:

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -126,39 +133,51 @@ spec = do
+        let poolMetrics = Types.ApiStakePoolMetrics (Quantity 0) (Quantity 0)
+        let pool = Types.ApiStakePool poolIdAbsent poolMetrics 0
+        w <- emptyWallet ctx
+        r <- joinStakePool ctx pool (w, "Secure Passphrase")
```
**[change required]**
`joinStakePool` requires a full `ApiStakePool` as argument which forces to create dummy values. I'd suggest to either:

1. drop `joinStakePool` and simply use the `request` combinator from the integration framework. 
2. Review `joinStakePool` and `joinStakePoolEp` to only require a stake pool id instead of a whole body

**Why**
The goal of the integration DSL is to favor readability and strive to have readable integration scenarios. By adding extra non-orthogonal (a.k.a overlapping) combinators, we do actually create noise and reduce readability and maintenance in the long run. Here, we are forced to create a fake pool metrics and a fake `ApiStakePool` object, whereas, we really just need to write:

```hs
    let poolId = T.replicate 64 "0"
    w <- emptyWallet ctx
    let payload = Json [aesonQQ| { "passphrase": "Secure Passphase" } |]
    request @(ApiTransaction 'Testnet) ctx (joinStakePoolEp poolId w) Default payload
```

paweljakubas replied:
I have spotted this, and planned to review it in forthcoming PR (here just used the same way as in other examples in the test suite). Added a point in development plan : 
```
[ ] review joinStakePoolEp in DSL
```
to tackle this in #896

paweljakubas committed:
comply with review of undrafted PR

remove comment

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +982,68 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+submitCert ctx wid (tx, meta, binary) = db & \DBLayer{..} -> do
+    withExceptT ErrSubmitCertNetwork $ postTx nw binary
+    mapExceptT atomically $ withExceptT ErrSubmitCertNoSuchWallet $ do
+        putTxHistory (PrimaryKey wid) [(tx, meta)]
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -786,15 +801,53 @@ listPools spl =
+    selection <- liftHandler $ withWorkerCtx ctx wid liftE1 $ \wrk ->
+        W.createCert @_ @s @t wrk wid
+
+    (tx, meta, time, wit) <- liftHandler $ withWorkerCtx ctx wid liftE2 $ \wrk ->
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -786,15 +801,53 @@ listPools spl =
+
+    allPools <- listPools spl
+    when (poolIsUnknown allPools) $
+        liftHandler $ throwE (ErrCreateCertNoSuchPool poolId)
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1579,6 +1635,43 @@ instance LiftHandler ErrMetricsInconsistency where
+                , toText poolId
+                ]
+        ErrCreateCertPoolAlreadyJoined poolId ->
+            apiError err403 PoolAlreadyJoined $ mconcat
```
:+1:

KtorZ reviewed (approved):
LGTM.

Perhaps a bit of cleanup / housekeeping the commit history before merging.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -110,6 +110,14 @@ module Cardano.Wallet
+    , createCert
+    , signCert
+    , submitCert
+    , ErrCreateCert (..)
```
Same point as I made here https://github.com/input-output-hk/cardano-wallet/pull/1059#discussion_r349318171: 

Because there are multiple kinds of certificates in jormungandr I find these names a bit ambiguous.

What do you think about:
- `createDelegationCert`
- `signDelegationCert`
- `submitDelegationCert`
- `ErrCreateDelegationCert`

paweljakubas replied:
sound argument, I will address it in the signCert (or signDelegationCert) PR soon

KtorZ replied:
I agree that naming could be clearer but in the context of the wallet, the term certificate is non-ambiguous. Jörmungandr is out of the question at this level (as well as in the API). 

Yet, it wouldn't hurt to add `Delegation` to make things clearer indeed.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -969,6 +982,68 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+    let sel = CoinSelection [] [] []
+    withExceptT ErrCreateCertFee $ do
+        debug tr "Coins selected for delegation certificate after fee adjustment"
+            =<< adjustForFee (feeOpts tl (computeCertFee 1) (bp ^. #getFeePolicy)) utxo sel
```
👌

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -786,15 +801,53 @@ listPools spl =
+    when (walMeta ^. #delegation == W.Delegating poolId) $
+        liftHandler $ throwE (ErrCreateCertPoolAlreadyJoined poolId)
+
+    allPools <- listPools spl
```
I'm not sure if this is wise.
1. `listStakePools` is not a reliable way to list the current `[PoolId]` because it has to be in sync with its chain-follower.*
2. This is using the api-`listPools`. So the error-messages from that endpoint can be thrown here I presume. I had never realised it was possible to nest api-calls like this though — cool! — but I don't think we want it?

Could we use `NetworkLayer.stakeDistribution` instead?

paweljakubas replied:
oh, thanks for this remark. As this is not the only one PR in this task, I will address it in the signCert (or signDelegationCert) PR soon

KtorZ replied:
what about simply relying on `DB.readPoolProduction` instead to avoid unnecessary networks roundtrips?

Anviking replied:
> what about simply relying on DB.readPoolProduction instead to avoid unnecessary networks roundtrips?

In order for the `DB` to contain any stake-distributions (would be `DB.readStakeDistribution`), it does at least need to _begin_ following the chain correctly. In contrast to the wallet, it needs to (I think it was)`fetchTip >> fetchStakeDistribution >> fetchTip` without the tip changing, which is a point of failure  unnecessary to bring in here.

Doing that would entangle the stake-pool & wallet state in a way I am worried about.

The network roundtrip should be _really_ insignificant here, anyway?

Anviking reviewed (approved):
lgtm! **But**, I don't think we'd like to use `listPools` from `joinStakePool`.
